### PR TITLE
8332724: x86 MacroAssembler may over-align code

### PR DESCRIPTION
**Pull request by djelinski**
The methods align32 and align64 are supposed to align the next instruction to the next 32 or 64 byte boundary using the minimum number of NOP bytes. However, when the target represented as a 32bit signed int is negative, the instructions generate 32 or 64 NOP bytes too many. This was observed in `jbyte_disjoint_arraycopy_avx3` on a Linux machine, where a single align32 invocation generated 63 bytes of NOPs.

This PR addresses the problem by using bit operations to calculate the required number of bytes.

Tier1-3 tests passed.

On a side note, `align64` and `align32` instructions were meant for aligning data for use with zmm / ymm loads, but nowadays they are frequently used in places where `align(CodeEntryAlignment)` or `align(OptoLoopAlignment)` would be more appropriate. I can address that in a separate PR if you think it's worth fixing.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warning
&nbsp;⚠️ Found leading lowercase letter in issue title for `8332724: x86 MacroAssembler may over-align code`

### Issue
 * [JDK-8332724](https://bugs.openjdk.org/browse/JDK-8332724): x86 MacroAssembler may over-align code (**Bug** - P5)

### Reviewers
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to [d0220193](https://git.openjdk.org/jdk/pull/19353/files/d0220193538619b8f42c21c5bac70fff59547eeb)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19353/head:pull/19353` \
`$ git checkout pull/19353`

Update a local copy of the PR: \
`$ git checkout pull/19353` \
`$ git pull https://git.openjdk.org/jdk.git pull/19353/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19353`

View PR using the GUI difftool: \
`$ git pr show -t 19353`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19353.diff">https://git.openjdk.org/jdk/pull/19353.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19353#issuecomment-2125698383)

### CODE REVIEW

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 1]
```diff
@@ -1164,8 +1164,13 @@ void MacroAssembler::align(int modulus) {
 }
 
 void MacroAssembler::align(int modulus, int target) {
```
Comment (dean-long) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:1166]: How about making both parameters unsigned?
And callers could be changed to something like:
```
align(64, (uint)(uintptr_t)pc() & 63);

```

Comment (djelinski) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp]: Good idea. It also fixes a couple of conversion warnings. Updated.

**Review by dean-long — APPROVED** [review]
Looks good.

**Comment by vnkozlov** [comment]
GHA x86-32 build issue:
```
/work/jdk/jdk/src/hotspot/cpu/x86/macroAssembler_x86.cpp: In member function ‘void MacroAssembler::align(uint)’:
/work/jdk/jdk/src/hotspot/cpu/x86/macroAssembler_x86.cpp:1162:18: error: comparison of integer expressions of different signedness: ‘uint’ {aka ‘unsigned int’} and ‘intx’ {aka ‘int’} [-Werror=sign-compare]
 1162 |   assert(modulus <= CodeEntryAlignment, "Alignment must be <= CodeEntryAlignment");
      |          ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
```

**Comment by djelinski** [comment]
Thanks for pointing it out! Should be fixed now, will wait for GHA to confirm...

**Review by vnkozlov — APPROVED** [review]
Good.

We need to revisit types of flags which hold small values. `intx` and `uintx` are 64-bits values in 64-bits VM.

**Comment by djelinski** [comment]
Thanks for the reviews. Unless someone objects, I'll integrate this later today.

Re revisiting flag types, it seems that the only options currently available are bool, int, intx and uintx, other types might [make jvmci unhappy](https://github.com/openjdk/jdk/blob/af056c1676dab3b0b35666a8259db60f9bbf824e/src/hotspot/share/jvmci/jvmciCompilerToVMInit.cpp#L294).

**Pull request closed without merging**